docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,423 @@
+- No requirement for users to have ASO knowledge
+- Greater freedom to change API implementations which we've recently leveraged to transition between the older
+  and newer Azure SDKs and to ASO.
+
+#### Option 6: Generate CAPZ code equivalent to what's added manually today
+
+For this option, a new code generation pipeline would be created to automatically scaffold the code that is
+currently manually written to expose additional AKS API fields to CAPZ.
+
+Once implemented, this method would drastically reduce the amount of developer effort to get started adding
+new AKS features. There would continue to be some amount of ongoing cost though to identify and handle nuances
+for each feature though, which may include issues outside of CAPZ's control like AKS API quirks.
+
+The main drawback of this approach from a developer perspective is the up-front effort to implement and
+ongoing cost to maintain the code generation itself. The existing AKS feature set exposed by CAPZ provides a
+decent foundation that can help identify regressions by testing the pipeline against features already
+implemented and tested. ASO also already implements a full code generation pipeline to transform Azure API
+specs into Kubernetes resource definitions, so some of that could possibly be reused by CAPZ.
+
+#### Option 7: CAPZ resource defines patches to ASO resource
+
+This method describes adding a new `spec.asoManagedClusterPatches` field to the existing
+AzureManagedControlPlane:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+
+	// ASOManagedClusterPatches defines patches to be applied to the generated ASO ManagedCluster resource.
+	ASOManagedClusterPatches []string `json:"asoManagedClusterPatches,omitempty"`
+}
+```
+
+After CAPZ calculates the ASO ManagedCluster for a given AzureManagedControlPlane during a reconciliation, it
+will apply these patches in order to the ManagedCluster before ultimately submitting that to ASO. This allows
+users to specify any AKS feature declaratively within the existing CAPZ spec. The exact format of the patches
+is TBD, but likely one or all of the formats described here:
+https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/
+
+The main drawback of this approach is the fragility of the patches themselves, which can break underneath
+users with a new version of ASO or CAPZ. There is also increased risk that a user could modify the ASO
+resource in a way that breaks CAPZ's ability to reconcile it, though CAPZ could perhaps add some extra
+validation that particularly sensitive fields do not get modified by a set of patches. Compared to other
+options, the syntax for specifying a patch is also more cumbersome than if its equivalent had its own CAPZ API
+field. Given the intent of setting patches is to enable niche use-cases for advanced users, these drawbacks
+may be acceptable.
+
+#### Option 8: Users bring-their-own ASO ManagedCluster resource
```

Comment:
I'm in favor of moving forward with this option (Option 8: Users bring-their-own ASO ManagedCluster resource) for the following reasons:

- Engineering cost is low.
  - Risk of embarking upon an effort that takes a long time, or in the worst case scenario, is never actually completed, is basically zero.
  - Risk of delivering a feature that users choose not to adopt is minimal.
    - If in fact the feature is rejected by the user community, we can drop without having sunk significant investment, and pivot to another solution.
- Enables users (at the cost of increased complexity to cluster template maintenance) to use *any* published AKS features in the ASO API that CAPZ supports. This reduces adoption time availability for almost all not-yet-implemented, or new, AKS features considerably.
- May actually accelerate new CAPZ development stories by enabling rapid functional prototyping of CAPZ + AKS feature work.

Would love to get @mtougeron's or @zmalik's or @dthorsen's or @LochanRn's thoughts here on whether or not this capability would actually solve any concrete problems.

Reply:
I just tried this out for myself and it's clunky enough that I think it would be hard to call a solution to this overall problem.

The main issue is that CAPI decouples creating worker nodes from control plane infra, but AKS requires node pools to be specified when the cluster is created. This means that to perform this Option 8, a user has to simulate what CAPZ does and create an ASO ManagedCluster with some `spec.agentPoolProfiles`, wait for the cluster to be created, then remove all of the agent pool specs from the ManagedCluster so as not to interfere with the AzureManagedMachinePool. To me, it doesn't seem possible for this option to achieve the main goal of allowing users to specify non-standard (to CAPZ) AKS features in a template and install that in one gesture. This is being addressed by AKS and tracked by ASO here: https://github.com/Azure/azure-service-operator/issues/2791. 

Another issue I noticed is that CAPZ currently doesn't report back on status for unmanaged ASO resources, so all user-created ASO resources are always assumed to be in a successful state. That by itself is simple enough to fix, though if we commit to that I think we would need to invent a new mechanism that users must employ to enable #4339 (like an `owner-azuremanagedcontrolplane` annotation). Overall I wonder if this option describes putting the ASO resources deep enough in a gray area between "inside" a CAPI cluster and "outside" one where it becomes hard to know exactly what the expectations on CAPZ are w.r.t. reading and modifying ASO resources.

Reply:
ACK 💭

Reply:
I don't have any ETA on this but I suspect that @phealy would welcome pushes on this:
> This is being addressed by AKS and tracked by ASO here: https://github.com/Azure/azure-service-operator/issues/2791.

Reply:
> I don't have any ETA on this but I suspect that @phealy would welcome pushes on this:

Even outside this specific context that would definitely smooth things out for CAPZ, so consider this a polite nudge on that 😄